### PR TITLE
Refactor code for moving from Herokuk to AWS lambda

### PR DESCRIPTION
**Pull request by ralren**
## [Get affinity bot into AWS](https://trello.com/c/eeYw9txo)

Changes proposed in this pull request:

- remove all references to flask and creating an app
- add in a dockerfile to package everything for deployment
- create a gha that will deploy the docker image to ecr for every push to main
- create a handler function as that is what lambda expects which will take all events

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ jobs:
       - name: Poetry Install
         uses: actions/setup-python@v4
         with:
-          python-version: "3.11.2"
+          python-version: "3.9"
```
Comment (ferlatte) [inline .github/workflows/test.yml:19]: Not a blocker and I'm curious why you're using Python 3.9 here over 3.10? Seems like using the latest stable release would be what I'd expect, and I'm sure I'm missing important details.

Comment (ferlatte) [reply .github/workflows/test.yml:19]: Apologies. I meant to type Python 3.11!

Comment (ralren) [reply .github/workflows/test.yml:19]: AWS Lambda can only support up to Python 3.9

File: .github/workflows/aws-ecr.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+          role-to-assume: ${{ secrets.GHA_ECR_ROLE_ASSUMPTION }}
+
+      - name: Login to Amazon ECR
+        id: login-ecr
+        uses: aws-actions/amazon-ecr-login@9149ade017c57f86dea2f76a01f8b2d5bd06b10f # tag=v1
+
+      - name: Login to ECR
+        uses: docker/login-action@49ed152c8eca782a232dede0303416e8f356c37b # tag=v2
+        with:
+          registry: ${{ steps.login-ecr.outputs.registry }}
```
Comment (jsclarridge) [inline .github/workflows/aws-ecr.yml:32]:
```suggestion
```
I don't think you need both `docker/login-action` and `aws-actions/amazon-ecr-login`. I think you can use either, but only one is needed.

**Comment by ferlatte** [comment]
Ok cool. Maybe add a comment in one of these to explain to future engineers?MOn Apr 13, 2023, at 3:59 PM, Ren ***@***.***> wrote:﻿
@ralren commented on this pull request.



In .github/workflows/test.yml:
> @@ -16,7 +16,7 @@ jobs:
       - name: Poetry Install
         uses: ***@***.***
         with:
-          python-version: "3.11.2"
+          python-version: "3.9"

AWS Lambda can only support up to Python 3.9

File: .github/workflows/aws-ecr.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    branches: [main]
+permissions:
+  id-token: write
+  contents: read
+
+jobs:
+  build-and-push-dev-aws:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b # tag=v3
```
[the remark this answers is not in the document]
Comment (rpdelaney) [reply .github/workflows/aws-ecr.yml:17]: Why not use the version tags? e.g.
```yaml
- uses: actions/checkout@v3
```
Also for the other actions.

File: pyproject.toml [hunk 4]
```diff
@@ -6,7 +6,7 @@ version = "0.1.0"
 readme = "README.md"
 
 [tool.poetry.dependencies]
-python = "^3.11"
+python = "^3.9"
 slackclient = "^2.9.4"
 flask = "^2.2.3"
```
Comment (rpdelaney) [inline pyproject.toml:11]:
```suggestion
```

Comment (rpdelaney) [reply pyproject.toml:11]: Also `pytest-flask` on L16

**Review by rpdelaney — APPROVED** [review]
It would be nice to have some of the CloudWatch logging stuff for debugging. But this is a big step in the right direction.

I think we just need to settle Josh's comments on the GHA stuff and run pre-commit hooks, then this is good to go I think :)

File: .github/workflows/test.yml (hunk 1 above)
Comment (rpdelaney) [reply .github/workflows/test.yml:19]: Mark's reply ended up getting added to the main thread: https://github.com/trussworks/affinity-groups-slackbot/pull/23#issuecomment-1507792845

> Ok cool. Maybe add a comment in one of these to explain to future engineers?

File: .github/workflows/aws-ecr.yml [hunk 5]
```diff
@@ -0,0 +1,67 @@
+name: Automated Docker Build Main Push AWS
+
+on:
+  push:
```
[the remark this answers is not in the document]
Comment (ralren) [reply .github/workflows/aws-ecr.yml:4]: Oh, I've already been testing by just pushing to this branch initially. Now that I'm pretty much done, I switched it back to main.

File: .github/workflows/aws-ecr.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ralren) [reply .github/workflows/aws-ecr.yml:17]: There is a reason to stick with GHA shas but I can't remember why. We should write an ADR to throw in legendary-waddle explaining why.

File: .github/workflows/aws-ecr.yml (hunk 2 above)
Comment (ralren) [reply .github/workflows/aws-ecr.yml:32]: Ah, good catch. I'll remove docker then.

File: .github/workflows/aws-ecr.yml [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        with:
+          registry: ${{ steps.login-ecr.outputs.registry }}
+
+      - name: Construct image URI
+        id: construct-image-uri
+        env:
+          GIT_BRANCH: ${{ github.ref_name }}
+          GIT_COMMIT: ${{ github.sha }}
+          ECR_REGISTRY: ${{ steps.login-ecr.outputs.registry }}
+          DEPLOY_ENV: ${{ inputs.deploy-env }}
```
[the remark this answers is not in the document]
Comment (ralren) [reply .github/workflows/aws-ecr.yml:40]: Another good catch.

File: .github/workflows/aws-ecr.yml [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        with:
+          install: true
+
+      - name: Build and push
+        uses: docker/build-push-action@3b5e8027fcad23fda98b2e3ac259d8d67585f671 # tag=v3
+        with:
+          context: .
+          push: true
+          tags: ${{ steps.construct-image-uri.outputs.image-uri }}
+          labels: ${{ steps.meta.outputs.labels }}
```
[the remark this answers is not in the document]
Comment (ralren) [reply .github/workflows/aws-ecr.yml:64]: and another

File: .github/workflows/aws-ecr.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsclarridge) [reply .github/workflows/aws-ecr.yml:17]: This doc goes into the context around using tags vs. shas with GHA: https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions

File: .github/workflows/test.yml (hunk 1 above)
Comment (ralren) [reply .github/workflows/test.yml:19]: Thanks for floating that up, Ryan!
I'll add a comment, Mark.

File: .github/workflows/aws-ecr.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsclarridge) [reply .github/workflows/aws-ecr.yml:4]: Ah, yeah. I should have looked through the commit history more carefully. I thought GHA would error out on the references to non-existent inputs or steps, but maybe GHA handles that gracefully or silently ignores them.

**Review by jsclarridge — APPROVED** [review]

File: .github/workflows/aws-ecr.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (rpdelaney) [reply .github/workflows/aws-ecr.yml]: I'm happy to land this as-is. I started a :lock:[slack thread](https://trussworks.slack.com/archives/CLNC1MUBS/p1681832582035079) so we can discuss how to do it going forward :)

**Pull request merged**